### PR TITLE
Qa convert

### PR DESCRIPTION
**Pull request by PsYear**

### CODE REVIEW

**Comment by PsYear** [comment]
—— tensorrt 使用的文档-----
保存好的 docker-image ：b9302a35bf50 - wendell-tensorrt-qa:latest 
可以直接运行的 docker-container：64e2ef06b4c7 - nvcr.io/nvidia/tensorrt:19.05-py3-installed-generate
直接attach上的命令:sudo nvidia-docker attach 64e2ef06b4c7
保持container运行的退出container方式:ctrl-p ctrl-q 
1.finetune的ktq任务名 130k-dianxiaomi-multi-128-pair-mix1648-shuffle-514-es-v0
2.将finetune的ckpt转换成bert源码里模型结构的ckpt （这一步应该可以省去，因为finetune的模型里，有一些处理n选一的结构，当时怕这部分op会影响转化，所以就用原来bert的模型结构再重新存了一遍，实际上 这个demo里tensorrt的转换方式是和op无关的，只和variable有关 ）
    Repo:n3lu    
    Branch:tensorrt_ckpt_generate
    Commit-id:8c550828e5bf0b92faf4d5c874fe845f5c5e9bd9
    运行参数：python -m bert.run_classifier --task_name=sim --do_train=true --do_eval=false --do_predict=false --vocab_file=/mnt/cephfs/yingkai/data/bert/dict/130k.txt --data_dir=/mnt/cephfs/wendell/data/similarity_task/re_build_question_10keys_40k_subtitle_cut/ctb_seg --predict_file=/mnt/cephfs/wendell/data/dureader/last_year/data_select_5/devset/search.dev.json --bert_config_file=/mnt/cephfs/yingkai/data/bert/config/130k.json --init_checkpoint=/mnt/cephfs/ktqueue/jobs/130k-dianxiaomi-multi-128-pair-mix1648-shuffle-514-es-v0/code/save/best-dev-loss-29193 --max_seq_length=128 --train_batch_size=6 --learning_rate=3e-5 --num_train_epochs=4.0 --output_dir=./save/ --do_single_char=false --num_gpus=1 --weight_decay=0.01 --use_adamw=True --customized_estimator=True --do_chinese_char=false --dev_start_epoch=0 --dev_rate_per_epoch=0.3 --debug_mode=true
    转换好的ckpt在save文件夹下,名字为ckpt-0-0
3. 将相关的文件上传到docker环境内，现在都放在docker的 /workspace/qa-ckpt/ 路径下
4.tensorrt docker里用到的code：https://github.com/naturali/TensorRT/tree/release/5.1 ，现在都放在docker的/workspace/TensorRT路径下
5.将ckpt转换成步骤7中tensorrt的weight。在文件 python/convert_weights.py中 param_names = … 为需要使用到的前馈variable，这里需要确保传入的variable必须要在模型计算中使用到，不然可能会在步骤7中出现错误。
运行路径: /workspace/TensorRT/demo/BERT
运行指令：python python/convert_weights.py -m /workspace/qa-ckpt/ckpt-0-0 -o tensorrt_convert/bert
6.将输入转换成步骤7中tensorrt读入的input。在文件 python/generate_dbg.py 中还会计算tensorflow的运算结果，用来做tensorrt输出的验证。输入和输出分别存为 test_inputs.weights_int32 和 test_outputs.weights，其中在计算中只会用到test_inputs.weights_int32，test_outputs.weights并不必须，只是用来做验证。 这里做了一些修改，加入了文本query传入时的预处理（tokenizer）,这里分词暂时用的是jieba分词，需要后续改成涛哥的分词。 此外因为会计算tensorflow的运算结果，所以为了保证结果和finetune的时候一致需要搭建一个和finetune时一样的模型结构（这里因为用了一个中间模型转换，所以是和中间模型一样，一般而言修改final_hidden=…之后 到logits 部分的过程 就可以了。）
运行路径: /workspace/TensorRT/demo/BERT
运行指令: python python/generate_dbg.py -f /workspace/qa-ckpt/ckpt-0-0 -p /workspace/qa-ckpt/ -o tensorrt_intput/ -s 128 -b 1 -d /workspace/qa-ckpt/130k.txt
7.运行tensorrt的runtime。在sampleBERT的代码里主要是两个部分，一个是生成bert的tensorrt计算图（engine=APIToModel（）…），另一个是计算结果(doInference)。在计算图中因为我们的模型是分类任务，所以写了一个专门处理分类任务的tensorrt的clf layer(ayers/clf.h)，这里主要是乘了pooler的weight并做tanh激活，以及再乘了得到logits的weight，另外还有原本就有squad的任务可以参考一下(layers/squad.h)。 在计算结果中，读入步骤6中的test_inputs.weights_int32 进行计算，并且比对了步骤6中test_outputs.weights的结果。
7-1 编译sampleBERT.cpp
运行路径 /workspace/TensorRT/demo/BERT/build
运行指令: cmake .. 
                make -j
7-2 执行 sampleBERT
运行路径
运行指令: ./sample_bert -d ./tensorrt_convert -d tensorrt_intput/  --nheads 12
Fp16 运行指令 ./sample_bert -d ./tensorrt_convert -d tensorrt_intput/  --fp16 --nheads 12
更进一步可以参考原来的demo：https://github.com/NVIDIA/TensorRT
